aptos-move/framework/aptos-framework/sources/staking_contract.move

```diff
@@ -0,0 +1,1132 @@
+///
+/// Example flow:
+/// 1. A staker creates a staking contract with an operator by calling create_staking_contract() with 100 coins of
+/// initial stake and commission = 10%. This means the operator will receive 10% of any accumulated rewards. A new stake
+/// pool will be created and hosted in a separate account that's controlled by the staking contract.
+/// 2. The operator sets up a validator node and, once ready, joins the validator set by calling stake::join_validator_set
+/// 3. After some time, the stake pool gains rewards and now has 150 coins.
+/// 4. Operator can now call request_commission. 10% of (150 - 100) = 5 coins will be unlocked from the stake pool. The
+/// staker's principal is now updated from 100 to 145 (150 coins - 5 coins of commission). The pending distribution pool
+/// has 5 coins total and the operator owns all 5 shares of it.
+/// 5. Some more time has passed. The pool now has 50 more coins in rewards and a total balance of 195. The operator
+/// calls request_commission again. Since the previous 5 coins have now become withdrawable, it'll be deposited into the
+/// operator's account first. Their new commission will be 10% of (195 coins - 145 principal) = 5 coins. Principal is
+/// updated to be 190 (195 - 5). Pending distribution pool has 5 coins and operator owns all 5 shares.
+/// 6. Staker calls unlock_stake to unlock 50 coins of stake, which gets added to the pending distribution pool. Based
+/// on shares math, staker will be owning 50 shares and operator still owns 5 shares of the 55-coin pending distribution
+/// pool.
+/// 7. Some time passes and the 55 coins become fully withdrawable from the stake pool. Due to accumulated rewards, the
+/// 55 coins become 70 coins. Calling distribute() distributes 6 coins to the operator and 64 coins to the validator.
+module aptos_framework::staking_contract {
+    use std::bcs;
+    use std::error;
+    use std::signer;
+    use std::vector;
+
+    use aptos_std::event::{EventHandle, emit_event};
+    use aptos_std::pool_u64::{Self, Pool};
+    use aptos_std::simple_map::{Self, SimpleMap};
+
+    use aptos_framework::account::{Self, SignerCapability};
+    use aptos_framework::aptos_coin::AptosCoin;
+    use aptos_framework::coin::{Self, Coin};
+    use aptos_framework::stake::{Self, OwnerCapability};
+    use aptos_framework::staking_config;
+
+    const SALT: vector<u8> = b"aptos_framework::staking_contract";
+
+    /// Store amount must be at least the min stake required for a stake pool to join the validator set.
+    const EINSUFFICIENT_STAKE_AMOUNT: u64 = 1;
+    /// Commission percentage has to be between 0 and 100.
+    const EINVALID_COMMISSION_PERCENTAGE: u64 = 2;
+    /// Staker has no staking contracts.
+    const ENO_STAKING_CONTRACT_FOUND: u64 = 3;
+    /// No staking contract between the staker and operator found.
+    const ESTAKING_CONTRACT_NOT_FOUND: u64 = 4;
```

Comment:
I wonder if these two should be "NO_STAKING_CONTRACT_FOUND_FOR_STAKER" and ""NO_STAKING_CONTRACT_FOUND_FOR_OPERATOR"?, current naming feels the same.

Reply:
Updated